### PR TITLE
chore: update nix to 2.17

### PR DESCRIPTION
**Pull request by tomberek**
brings Nix to 2.17

- brought in some patches to make nlohmann machinery work, ~~notably adding a variant<A,B,C> handler which is a bit ugly, but generic handling of arbitrary sum types in C++ is....... an exercise for the reader.~~ restored support for sum types
- included the multiple-github-tokens patch
- use overlays to make `nix` the patched version for all downstream users of the pkgdb overlays.

### CODE REVIEW

**Comment by aakropotkin** [comment]
I think we want to override `nix` in our overlay so that anyone else in the closure using `nix` follows our version.

Do you still have your old PR we wrote for 2.17?

File: include/flox/core/nix-state.hh [hunk 1]
```diff
@@ -139,15 +141,16 @@ public:
...
   getState()
   {
     if ( this->state == nullptr )
       {
-        this->state
-          = std::make_shared<nix::EvalState>( std::list<std::string> {},
-                                              this->getStore(),
-                                              this->getStore() );
+        this->state = std::make_shared<nix::EvalState>(
+          nix::SearchPath()
```
Comment (mkenigs) [inline include/flox/core/nix-state.hh:150]: question nonblocking: why this change?

**Review by mkenigs — APPROVED** [review]

File: include/flox/core/nix-state.hh (hunk 1 above)
Comment (tomberek) [reply include/flox/core/nix-state.hh:150]: upstream changed the API for this function call. (ref: https://github.com/NixOS/nix/pull/8579/files#diff-f118e4c6f6e02148b887fdf627352311fca5a3a4eadf0b4a9d9f348e0be464ffL500-L501)
